### PR TITLE
6935 remove order string syntax

### PR DESCRIPTION
**Pull request by mkaufmaner**
_Thanks for wanting to fix something on Sequelize - we already love you long time! Please delete this text and fill in the template below. If unsure about something, just do as best as you're able._

_If your PR only contains changes to documentation, you may skip the template below._

### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Have you added an entry under `Future` in the changelog?

_NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._

### Description of change
Closes https://github.com/sequelize/sequelize/issues/6935

### CODE REVIEW

**Comment by mkaufmaner** [comment]
Closed and reopened to trigger builds from commit that resolves change log conflict.

**Comment by janmeier** [comment]
LGTM, would like one more @sequelize/core member to look at it

File: lib/dialects/abstract/query-generator.js [hunk 1]
```diff
@@ -1483,24 +1483,40 @@ const QueryGenerator = {
...
       }
     };
 
+    const getOrderOption = order => {
+      const upperOrder = order.toUpperCase();
+      let index = -1;
+      let option = false;
+
+      for (const v of validOrderOptions) {
+        if(_.endsWith(upperOrder, v)){
```
Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: space after `if` and before opening brace

File: lib/dialects/abstract/query-generator.js [hunk 2]
```diff
@@ -1519,8 +1535,26 @@ const QueryGenerator = {
...
+        order = _.trim(order);
+        let sql = this.quote(order, model, false, '->');
+        if(orderOption.option){
+          sql += ' ' + orderOption.option;
+        }
+        if (subQuery) {
+          subQueryOrder.push(sql);
+        }
+        mainQueryOrder.push(sql);
+      }
```
Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: Wait, how is this removing the string order option? Passing a string to `order` should throw an exception. If you want to order by a literal, use `Sequelize.literal()`. Any custom sanitisation is prone to be spoofed.

File: changelog.md [hunk 3]
```diff
@@ -1,4 +1,5 @@
 # Future
+- [FIXED] Removed `Sequelize.literal()` support for order strings [#6935](https://github.com/sequelize/sequelize/issues/6935)
```
Comment (felixfbecker) [inline changelog.md]: This is the exact opposite of what we want. `Sequelize.literal()` must still be supported, only passing a string should be removed.

File: lib/dialects/abstract/query-generator.js (hunk 1 above)
Comment (mkaufmaner) [inline lib/dialects/abstract/query-generator.js]: Fixed.

File: changelog.md (hunk 3 above)
Comment (mkaufmaner) [inline changelog.md]: Fixed https://github.com/sequelize/sequelize/pull/7160/commits/33fa62050fc81f9b90b337847a78078d96443b68#diff-42ba1d994f4fa8c6ad17a7efae7936cc

File: lib/dialects/abstract/query-generator.js (hunk 2 above)
Comment (mkaufmaner) [inline lib/dialects/abstract/query-generator.js]: @felixfbecker The string order option is not removed and you can still use `Sequelize.literal()`.  However, we can continue to support strings without breaking existing functionality.  Sure, custom sanitation methods are prone to being spoofed, however the order clause has a simple syntax where we can easily validate content and its arrangement.  So why loose existing functionality and when it could be extended?  If worst comes to worst, the logic being implemented here will quote a malicious or malformed order string and error out.

Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: We don't support raw strings in other places in favor of explicit Sequelize.literal(), like `attributes` etc. `order` is one of the last places where this is still supported. Sequelize codebase is already very complicated, there is also the desire to keep complexity low. Can you really guarantee that your algorithm can't be spoofed? Can you guarantee you handle every possible ORDER syntax? Just look at the possibilities for Postgres: https://www.postgresql.org/docs/9.1/static/queries-order.html, these also can vary through dialects. Since we don't test all the edge cases here there is still the potential to break someone's code and I don't want to commit to parsing every case here just to then put it inside the SQL again. Essentially this would introduce a mini DSL for `order` that is probably a subset of what's possible with SQL and it would still be an attack vector for SQL injection. I would rather target v4 and tell people to wrap in Sequelize.literal().

Comment (mkaufmaner) [inline lib/dialects/abstract/query-generator.js]: @felixfbecker Understood.  Will modify the tests/logic/documentation to only accept array.  Then I will do `group` option then.

Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: Awesome :)

File: lib/dialects/abstract/query-generator.js (hunk 1 above)
Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: uhm, no?

**Comment by sushantdhiman** [comment]
This is a breaking change in `order` api, as now we do not allow passing strings. Need to mention this in breaking changes section for Future release

File: lib/dialects/abstract/query-generator.js [hunk 4]
```diff
@@ -1520,13 +1521,16 @@ const QueryGenerator = {
 
         mainQueryOrder.push(this.quote(t, model, false, '->'));
       }
-    } else {
-      var sql = this.quote(typeof options.order === 'string' ? new Utils.Literal(options.order) : options.order, model, false, '->');
+    } else if (typeof options.order === 'function' && options.order instanceof this.sequelize.literal) {
```
Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: If it is an `instanceof` a literal, shouldn't it be `typeof 'object'`?

File: test/unit/dialects/sqlite/query-generator.test.js [hunk 5]
```diff
@@ -168,8 +168,12 @@ if (dialect === 'sqlite') {
           expectation: 'SELECT count(*) AS `count` FROM `foo`;',
           context: QueryGenerator
         }, {
+          arguments: ['myTable', {order: 'id; DROP TABLE myTable;'}],
+          expectation: 'SELECT * FROM `myTable` ORDER BY `id; DROP TABLE myTable;`;',
```
Comment (felixfbecker) [inline test/unit/dialects/sqlite/query-generator.test.js]: This is what should throw an exception instead

File: lib/dialects/abstract/query-generator.js (hunk 4 above)
Comment (mkaufmaner) [inline lib/dialects/abstract/query-generator.js]: @felixfbecker There are a bunch of changes that need to be done as well.  These are just temporary.  Can you change the status of this PR to pending?

Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: What do you mean with changing the status to pending?

Comment (mkaufmaner) [inline lib/dialects/abstract/query-generator.js]: Sorry, just woke up.  I mean remove the label "pr:final review"

**Comment by mkaufmaner** [comment]
@felixfbecker Looks good!  Please review changes and let me know.  Take not of https://github.com/sequelize/sequelize/pull/7160/commits/3bd6aa118c4b47a8c8c99d51b3ae9fe265246908

**Comment by mkaufmaner** [comment]
@sushantdhiman "Not completed yet"?

**Comment by felixfbecker** [comment]
@mkaufmaner 
He just dismissed his approval.

If you want to trigger a build, please just `git commit --amend --no-edit; git push --force`, reopening sends emails to everyone in the thread

**Comment by mkaufmaner** [comment]
@felixfbecker 
Gotcha and sorry about the multiple emails.  Again, not familiar with team development on github.
`git commit --amend --no-edit; git push --force` it is.

Please review and comment :)

**Comment by felixfbecker** [comment]
@mkaufmaner And of course, if you just pushed a regular commit, that will trigger a build too.

File: changelog.md [hunk 6]
```diff
@@ -48,6 +49,7 @@
 ## BC breaks:
 - `DATEONLY` now returns string in `YYYY-MM-DD` format rather than `Date` type
 - With `BelongsToMany` relationships `add/set/create` setters now set `through` attributes by passing them as `options.through` (previously second argument was used as `through` attributes, now its considered `options` with `through` being a sub option)
+- `options.order` Now only excepts values with type of array or sequelize method.  Support for string values (ie `{order: 'name DESC'}`) has been deprecated.
```
Comment (felixfbecker) [inline changelog.md]: ~~deprecated~~ removed

File: lib/dialects/abstract/query-generator.js [hunk 7]
```diff
@@ -21,7 +20,7 @@ const QueryGenerator = {
     tableName = tableName || {};
     return {
       schema: tableName.schema || options.schema || 'public',
-      tableName: _.isPlainObject(tableName) ? tableName.tableName : tableName,
+      tableName: Utils._.isPlainObject(tableName) ? tableName.tableName : tableName,
```
Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: We moved away from using `Utils._`. This makes the diff unnecessary large.

File: lib/dialects/abstract/query-generator.js [hunk 8]
```diff
@@ -1484,48 +1483,68 @@ const QueryGenerator = {
...
+      'ASC NULLS LAST',
+      'DESC NULLS LAST',
+      'ASC NULLS FIRST',
+      'DESC NULLS FIRST',
+      'NULLS FIRST',
+      'NULLS LAST'
+    ];
+
+    const methodOrder = order => {
+      let result = false;
```
Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: you don't need this variable

File: lib/dialects/abstract/query-generator.js [hunk 9]
```diff
@@ -1484,48 +1483,68 @@ const QueryGenerator = {
...
+      'DESC',
+      'ASC NULLS LAST',
+      'DESC NULLS LAST',
+      'ASC NULLS FIRST',
+      'DESC NULLS FIRST',
+      'NULLS FIRST',
+      'NULLS LAST'
+    ];
+
+    const methodOrder = order => {
```
Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: this function should probably be in `utils` and called `isSequelizeMethod`

File: lib/dialects/abstract/query-generator.js [hunk 10]
```diff
@@ -1484,48 +1483,68 @@ const QueryGenerator = {
...
             }
           } else {
-            validateOrder(_.last(t));
+            validateOrder(Utils._.last(order));
           }
         }
 
-        if (subQuery && (Array.isArray(t) && !(typeof t[0] === 'function' && t[0].prototype instanceof Model) && !(t[0] && typeof t[0].model === 'function' && t[0].model.prototype instanceof Model))) {
-          subQueryOrder.push(this.quote(t, model, false, '->'));
+        if (subQuery && (Array.isArray(order) && !(Utils._.isFunction(order[0]) && order[0].prototype instanceof Model) && !(order[0] && Utils._.isFunction(order[0].model) && order[0].model.prototype instanceof Model))) {
```
Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: Why the change to lodash typechecking functions everywhere? `typeof` is way faster than lodash's cross-browser implementation.

File: lib/dialects/sqlite/query-generator.js [hunk 11]
```diff
@@ -332,7 +332,7 @@ const QueryGenerator = {
 
   quoteIdentifier(identifier) {
     if (identifier === '*') return identifier;
-    return Utils.addTicks(identifier, '`');
+    return Utils.addTicks(Utils.removeTicks(identifier, '`'), '`');
```
Comment (felixfbecker) [inline lib/dialects/sqlite/query-generator.js]: why is this needed?

File: changelog.md (hunk 6 above)
Comment (felixfbecker) [inline changelog.md]: capitalize Sequelize

File: lib/dialects/abstract/query-generator.js (hunk 7 above)
Comment (mkaufmaner) [inline lib/dialects/abstract/query-generator.js]: Okay, so can we move completely away from using `Utils._.lodashMethod` and  change everything to be `_.lodashMethod`?

File: changelog.md (hunk 6 above)
Comment (mkaufmaner) [inline changelog.md]: Done

File: lib/dialects/abstract/query-generator.js (hunk 10 above)
Comment (mkaufmaner) [inline lib/dialects/abstract/query-generator.js]: Okay, I will change it back.

File: lib/dialects/abstract/query-generator.js (hunk 7 above)
Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: yes. In many cases you don't even need lodash because theres a vanilla JS equivalent for it now

File: lib/dialects/abstract/query-generator.js (hunk 9 above)
Comment (mkaufmaner) [inline lib/dialects/abstract/query-generator.js]: Well, I think I'll just remove this all together in favor of just checking if the `_isSequelizeMethod` property is set.

Comment (felixfbecker) [inline lib/dialects/abstract/query-generator.js]: No, we need the `instanceof` check. `_isSequelizeMethod` can easily be spoofed from JSON input.

File: lib/dialects/sqlite/query-generator.js (hunk 11 above)
Comment (mkaufmaner) [inline lib/dialects/sqlite/query-generator.js]: This is probably one of the most important changes features that prevents injection.  Similar to how values are escaped, this escapes identifiers by removing them.  See line 580 of `postgres/query-generator.js`.

File: lib/dialects/abstract/query-generator.js (hunk 8 above)
Comment (mkaufmaner) [inline lib/dialects/abstract/query-generator.js]: I removed this so it is irrelevant now.

File: lib/dialects/sqlite/query-generator.js (hunk 11 above)
Comment (mkaufmaner) [inline lib/dialects/sqlite/query-generator.js]: @felixfbecker Maybe instead, this should escape the identifiers instead of removing them.  I think this is what `escapeId` on line 7 of `/lib/sql-string.js` was originally meant to do.

Comment (mkaufmaner) [inline lib/dialects/sqlite/query-generator.js]: @felixfbecker After more consideration, we should not escape identifiers and just remove them.  Supporting identifiers with identifying characters in them seems silly.

**Comment by janmeier** [comment]
Good work @mkaufmaner ! :tada:
